Adds AdeMAMix Optimizer to `contrib`

mathDR opened the pull request:
This PR adds the AdaMAMix optimizer from The arxiv preprint: [THE ADEMAMIX OPTIMIZER:
BETTER, FASTER, OLDER](https://arxiv.org/pdf/2409.03137)

Closes #1058 

The docs have been updated, along with the relevant files. 

Currently the docstrings are implemented, but further descriptions should/could be added.  I will reach out to the paper authors to assist with that (if they are willing).

zcharles8 commented on optax/contrib/_ademamix.py:
```diff
@@ -0,0 +1,182 @@
+  """State for the Ademamix algorithm."""
+
+  count: chex.Array
+  count_m2: chex.Array
```
I'm not sure why you have multiple counts. Would just the first suffice?

mathDR replied:
So the first `count` replicates the same entitiy in `adam`.  The second is for keeping track of the second momentum, so the user might want to do different things with each.

Does that make sense?

I will make an inline comment to this fact

zcharles8 commented on examples/contrib/rosenbrock_ademamix.ipynb:
Super cool work - reproducing results is definitely under-appreciated. This is really nice!

zcharles8 commented on optax/contrib/_ademamix.py:
```diff
@@ -0,0 +1,182 @@
+    m1 = otu.tree_update_moment(
+      updates, state.m1, b1, 1
+    )  # m1 = b1 * m1 + (1-b1) * updates
+    m2 = otu.tree_update_moment(updates, state.m2, c_b3, 1)
```
I know this isn't done elsewhere in optax, but generally I prefer opaque constants (e.g. 1 in this case) to be passed in via kwargs. Consider this optional though.

mathDR replied:
so:
```
m2 = otu.tree_update_moment(updates, state.m2, c_b3, order=1)
```
is what you are describing?

zcharles8 replied:
Exactly.

zcharles8 commented on optax/contrib/_ademamix.py:
```diff
@@ -0,0 +1,182 @@
+    nu = otu.tree_update_moment_per_elem_norm(updates, state.nu, b2, 2)
+    count_inc = numerics.safe_int32_increment(state.count)
+    count_m2_inc = numerics.safe_int32_increment(state.count_m2)
+    m1_hat = otu.tree_bias_correction(m1, b1, count_inc)
```
It could be useful to leave a comment here noting that AdEMAMix does not perform bias correction on m2.

mathDR replied:
Yes great point.  This is in addition to scheduling `b3` and `alpha` to alleviate early training instabilities.  I will make an inline note and elaborate the issue in the description.

mathDR commented on docs/api/contrib.rst:
```diff
@@ -82,6 +83,12 @@ Momo
 .. autofunction:: momo_adam
 .. autoclass:: MomoAdamState
 
+Multiple EMA AdEMAMix
```
Done

mathDR commented on optax/contrib/_ademamix.py:
```diff
@@ -0,0 +1,182 @@
+class ScaleByAdemamixState(NamedTuple):
+  """State for the Ademamix algorithm."""
+
+  count: chex.Array
```
Oh nice.  I see it is 
```
count: chex.Array  # shape=(), dtype=jnp.int32.
```
everywhere.

mathDR commented on optax/contrib/_ademamix.py:
```diff
@@ -0,0 +1,182 @@
+  b3: float = 0.9999,
+  alpha: float = 5.0,
+  b3_scheduler: Optional[base.ScalarOrSchedule] = None,
+  alpha_scheduler: Optional[base.ScalarOrSchedule] = None,
```
I noticed the patter with the majority of the other classes that did not use `base.ScalarOrSchedule` as the type and tried to replicate that.

Of course, the authors in the paper specifically recommend scheduling `b3` and `alpha` because of early training instabilities, so this is a bit of a different use case.

I will make this change.

mathDR commented:
Hey @zcharles8 in rereading the paper I subsequently discovered the authors have a full `jax` implemenentation [here](https://github.com/apple/ml-ademamix).

I will email them and see if they have any qualms about me using aspects of their code as part of this PR.

Do you know if there would be licensing problems with this approach?

Please advise.

vroulet commented:
> Do you know if there would be licensing problems with this approach?

Great question, probably there would from reading Apple's license... Let me send a mail to one of the authors I know.

mathDR commented:
Thanks @vroulet I also tried sending an email to all 3 authors (but had to google their contact info so those might be outdated)

Basically I just asked if they are keen to get a version into `contrib` I would continue my PR but use their docstrings (with attribution)

mathDR commented:
Emailed with Matteo Pagliardini and he wrote: 
```
Very happy to hear that you found the work useful and thanks for the PR. Having AdEMAMix as part of Optax would be great. Feel free to proceed with the PR and use our docstrings. 
```
So I went ahead and completed the PR.

One open question:  AdEMAMix uses a pretty bespoke scheduler for `b3`.  The `alpha` scheduler can be implemented via the vanilla `linear_schedule` but I couldn't find a drop in replacement for the `b3` scheduler.

Both are now used in the `rosenbrock` example, but I didn't know if we wanted to add a new scheduler type for completeness?

zcharles8 commented on optax/contrib/_ademamix.py:
```diff
@@ -1,39 +1,38 @@
 from optax._src import transform
-
+import optax.tree_utils as otu
+from typing import NamedTuple, Tuple
```
I think the convention in this library is that imports from python generally go at the top, so
```
from typing import...

import chex
...
```

mathDR replied:
Ah!  Good point.  (I was going to make a meta point about this repo not having any `pre-commit` hooks but...).

zcharles8 commented:
> One open question: AdEMAMix uses a pretty bespoke scheduler for `b3`. The `alpha` scheduler can be implemented via the vanilla `linear_schedule` but I couldn't find a drop in replacement for the `b3` scheduler.
> 
> Both are now used in the `rosenbrock` example, but I didn't know if we wanted to add a new scheduler type for completeness?

I think that adding the schedulers to the library (`alpha_scheduler` and `b3_scheduler`) is totally reasonable, if for no other reason than to make it easy for someone to directly use the schedulers from the paper.

FWIW I think that this PR is really good (ie. things like typing information) and would prefer to have it pushed through (nice work @mathDR !)

vroulet commented on optax/contrib/_ademamix.py:
```diff
@@ -0,0 +1,220 @@
+) -> base.GradientTransformation:
+  """Rescale updates according to the Ademamix algorithm.
+
+  References:
```
Use the rst format that is:
```
    Pagliardini et al, `The AdEMAMix Optimizer: Better, Faster, Older
    <https://arxiv.org/abs/2409.03137>`_, 2024
 ```

vroulet commented on optax/contrib/_ademamix.py:
```diff
@@ -0,0 +1,220 @@
+  eps: float = 1e-8,
+  eps_root: float = 0.0,
+) -> base.GradientTransformation:
+  """Rescale updates according to the Ademamix algorithm.
```
Could you update this docstring too? The `ademamix` docstring looks great. You may simply refer here to the docstring of `ademamix` for e.g. the pseudocode and copy-paste e.g. the descriptions of the arguments.

mathDR replied:
So to clarify:  just a link to the `ademamix` docstring?  That is, don't follow the convention of:
```
  Description

  References:

  Args:

  Returns:
    A `GradientTransformation` object.
```
?

Note:  the above convention is prevalent throughout `optax/_src/transform.py` for similar `scale_by_X` functions.

vroulet commented on optax/contrib/_ademamix.py:
```diff
@@ -0,0 +1,220 @@
+import optax.tree_utils as otu
+
+class ScaleByAdemamixState(NamedTuple):
+  """State for the Ademamix algorithm."""
```
Could you add a description of the attributes in the docstring?
See e.g. https://github.com/google-deepmind/optax/blob/b8c2e133319480509576a7280f851e9d6ec7dccf/optax/_src/transform.py#L1448

mathDR replied:
I took a stab.  LMK if there are better descriptions I could do.  I (theoretically) could add the math update if you believe that would be warranted.  I also want to get the description right.  `nu` is still the estimate of the second moment, but now the first moment is a combination of `m1` and `m2` so I just stated their respective EMA types.

vroulet replied:
Looks great to me.
I think I would say "fast EMA of first moment" rather than first moment personally (same for "slow EMA" and this would apply in the other docstrings). But if you have stronger preferences I won't argue against.

vroulet reviewed (commented):
Looks great. Thank you @mathDR ! And thank you @zcharles8 for the review!

I've left a few minor comments. Mostly, it would be good to clean up the docstring of scale_by_ademamix now that you did the docstring of ademamix. And you;ll be able to resolve any remaining comments that way.

Thank you again to both of you!

vroulet reviewed (commented):
OK, final comments and then we'll be good to go I think

mathDR commented:
Okay thanks for the tip to render the docs.  That allowed me to fix a _lot_ of weirdness (and a LaTeX error!).

I think this is good to go!

vroulet reviewed (approved):
Thank you again @mathDR !

mathDR commented:
So @vroulet I think you have to merge it?  Or does another authorized user have to do it?

fabianp commented:
thanks for the contribution! Note that you'll need to edit `gallery.rst` for your example to display in https://optax.readthedocs.io/en/latest/gallery.html

mathDR commented:
Okay great.  I updated `gallery.rst` and added a png thumbnail to the `images/` directory.  I also added a `colab` link, as I saw the other examples did the same.

Note, when I render the docs locally, my example renders a `Keyboard Interrupt` that doesn't exist in the original jupyter notebook.

Is this an artifact of sphinx attempting to render the notebook? Hopefully this is a problem with my local env and will not persist in `main`

fabianp commented:
that's because your test is taking too long to run. you can either make it quicker, or add it to nb_execution_excludepatterns in https://github.com/google-deepmind/optax/blob/main/docs/conf.py (but then it won't be run as part of the test suite)

fabianp commented:
i think tests are not passing?

fabianp commented:
I think it's mostly good, except some conflicts with optax/contrib/_common_test.py

mathDR commented:
The failing tests both seem to come from `lbfgs.py` wherein they error with:
```
**********************************************************************
File "../optax/_src/alias.py", line ?, in default
Failed example:
    for _ in range(5):
      value, grad = value_and_grad(params, state=opt_state)
      updates, opt_state = solver.update(
         grad, opt_state, params, value=value, grad=grad, value_fn=f
      )
      params = optax.apply_updates(params, updates)
      print('Objective function: ', f(params))
Expected:
    Objective function:  7.516686
    Objective function:  7.460699e-14
    Objective function:  2.6505726e-28
    Objective function:  0.0
    Objective function:  0.0
Got:
    Objective function:  7.5166864
    Objective function:  7.460699e-14
    Objective function:  2.6505726e-28
    Objective function:  0.0
    Objective function:  0.0
**********************************************************************
```
the apparently passes on `main` though, so IDK why it is failing now.

mathDR commented:
Okay IDK why the tests above fail.  It appears that `main` is passing all of the tests, but @fabianp can you confirm that your tests all pass (and do not replicate the error I see above)?  I am a bit stuck here on how to proceed:  I don't really want to touch any code in `../optax/_src/alias.py` since none of that is part of the PR...

rdyro commented:
> Okay IDK why the tests above fail. It appears that `main` is passing all of the tests, but @fabianp can you confirm that your tests all pass (and do not replicate the error I see above)? I am a bit stuck here on how to proceed: I don't really want to touch any code in `../optax/_src/alias.py` since none of that is part of the PR...

Hi @mathDR I'll try to take over on this to get this merged. This doctest is very sensitive to changes in numerics (it's not ideal), can you disable it in this PR please (in `alias.py`) like so:

`print('Objective function: ', f(params))  # doctest: +SKIP`

The doctests should then pass and we can merge

mathDR commented:
@rdyro I just put the SKIP on all of the print statements (probably overkill, but....)

All of the tests now pass! 🥳

rdyro commented:
> @rdyro I just put the SKIP on all of the print statements (probably overkill, but....)
> 
> All of the tests now pass! 🥳

Ah, noooooooo :D, we DO want to keep those doctests. 

I didn't realize earlier, but other doctests reduce precision for comparison. These are all deterministic tests, so a lower accuracy comparison is by far the best choice. Could you revert the doctest skips, but add formatting to the lbfgs test:

The line should be:
```
  print('Objective function: {:.2E}'.format(f(params)))
```

And the output should be (I just ran it manually for reference):
```
Objective function: 7.52E+00
Objective function: 7.46E-14
Objective function: 2.65E-28
Objective function: 0.00E+00
Objective function: 0.00E+00
```

Sorry for the confusion...

mathDR commented:
Oh great catch @rdyro !  I updated the code as requested.  The tests are currently running, but the doctests have passed.  I think we are good to go...

fabianp commented:
please add the following copyright to new files:

```
# Copyright 2025 DeepMind Technologies Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
```

mathDR commented:
Just pinging you here @fabianp since I _think_ we are all good?

rdyro commented:
> Just pinging you here @fabianp since I _think_ we are all good?

Looks great, I'm in the process of merging it internally, thanks for this contribution!
